src/librbd/journal/CreateJournalRequest.cc

```diff
@@ -0,0 +1,235 @@
+  m_journaler = new Journaler(m_op_work_queue, m_timer, m_timer_lock,
+                              m_ioctx, m_image_id, m_image_client_id,
+                              m_cct->_conf->rbd_journal_commit_age);
+
+  using klass = CreateJournalRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_create_journal>(this);
+
+  m_journaler->create(m_order, m_splay_width, m_pool_id, ctx);
+}
+
+template<typename I>
+Context *CreateJournalRequest<I>::handle_create_journal(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "failed to create journal: " << cpp_strerror(*result) << dendl;
+    shut_down_journaler(*result);
+    return nullptr;
+  }
+
+  allocate_journal_tag();
+  return nullptr;
+}
+
+template<typename I>
+void CreateJournalRequest<I>::allocate_journal_tag() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  using klass = CreateJournalRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_journal_tag>(this);
+
+  ::encode(m_tag_data, m_bl);
+  m_journaler->allocate_tag(m_tag_class, m_bl, &m_tag, ctx);
+}
+
+template<typename I>
+Context *CreateJournalRequest<I>::handle_journal_tag(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "failed to allocate tag: " << cpp_strerror(*result) << dendl;
+    shut_down_journaler(*result);
+    return nullptr;
+  }
+
+  register_client();
+  return nullptr;
+}
+
+template<typename I>
+void CreateJournalRequest<I>::register_client() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  m_bl.clear();
+  ::encode(ClientData{ImageClientMeta{m_tag.tag_class}}, m_bl);
+
+  using klass = CreateJournalRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_register_client>(this);
+
+  m_journaler->register_client(m_bl, ctx);
+}
+
+template<typename I>
+Context *CreateJournalRequest<I>::handle_register_client(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "failed to register client: " << cpp_strerror(*result) << dendl;
+  }
+
+  shut_down_journaler(*result);
+  return nullptr;
+}
+
+template<typename I>
+void CreateJournalRequest<I>::shut_down_journaler(int r) {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  m_r_saved = r;
+
+  using klass = CreateJournalRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_journaler_shutdown>(this);
+
+  m_journaler->shut_down(ctx);
+}
+
+template<typename I>
+Context *CreateJournalRequest<I>::handle_journaler_shutdown(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "failed to shut down journaler: " << cpp_strerror(*result) << dendl;
+  }
+
+  delete m_journaler;
+
+  if ((m_r_saved == 0) || (*result < 0)) {
```

Comment:
@dillaman Here, the cleanup cuts short if an error in encountered when shutting down journaler. Would it makes sense to continue deleting journal objects in this case?